{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9decd6be-e1b4-473b-97e0-5c1f31686c3e",
   "metadata": {
    "id": "9decd6be-e1b4-473b-97e0-5c1f31686c3e"
   },
   "source": [
    "# **Submodule 1.2 - Protein Structure Analysis**\n",
    "\n",
    "## **Learning Objectives:**\n",
    "1. Understand dihedral and torsion angles\n",
    "2. Understand bond angles\n",
    "3. Understand bond energy and distances\n",
    "4. Understand homology analysis\n",
    "\n",
    "## **Prerequisites:**\n",
    "**Conceptual Knowledge:**\n",
    "- **Protein Basics:** Familiarity with amino acids, peptide bonds, and the four levels of protein structure (primary, secondary, tertiary, quaternary).\n",
    "- **Dihedral Angles** Understanding of dihedral angles (phi, psi, chi) and their role in protein conformation and stability.\n",
    "\n",
    "**Software Proficiency:**\n",
    "- **PyMOL Basics:** Basic proficiency in using PyMOL for protein visualization, including loading PDB files, changing display modes, and navigating the interface.\n",
    "- **Python Fundamentals:** Basic understanding of Python programming, including syntax, variables, loops, and functions, for script-based analysis.\n",
    "\n",
    "## **Introduction**\n",
    "This submodule explores the fundamental aspects of protein geometry and evolutionary relationships that govern protein structure and function. Using PyMOL as our visualization platform, we'll investigate how the backbone dihedral angles (œÜ and œà) and side chain torsion angles define protein conformation, and how these parameters are visualized and analyzed using Ramachandran plots. Through hands-on exercises, you'll learn to distinguish between conformational and configurational changes, understand the allowed regions of protein geometry, and recognize how these geometric constraints influence protein folding and structure. Building on these geometric principles, we'll examine how protein sequences and structures are related through evolution, using homology analysis to understand conservation patterns and structural similarities across protein families. You'll learn to analyze sequence alignments, identify conserved structural motifs, and appreciate how this information can be applied to predict protein structures and guide drug design."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df290b07-14cc-4f5a-8947-762b01a5ad06",
   "metadata": {
    "id": "df290b07-14cc-4f5a-8947-762b01a5ad06"
   },
   "source": [
    "## **Protein Conformation and Dihedral Angles:**\n",
    "Protein conformation refers to the spatial arrangement of atoms in a protein, primarily determined by the rotation around single bonds in the polypeptide backbone. This flexibility allows proteins to adopt different shapes that are crucial for their biological functions. The dihedral angles‚Äîdenoted as Œ¶ (phi) and Œ® (psi) ‚Äîare the key rotational parameters in proteins. These angles represent the rotations between the nitrogen and alpha carbon (Œ¶) and between the alpha carbon and carbonyl carbon (Œ®). The IUPAC definition of dihedral angles helps standardize how these rotations are described in the context of protein structures.\n",
    "\n",
    "<center><img src=\"images/dihedral_angles.png\" width=800 /></center><br><br>\n",
    "\n",
    "The specific values of Œ¶ and Œ® angles influence the secondary structure that a protein can adopt (e.g., alpha-helix, beta-sheet). Dihedral angles also affect the energy states of proteins, where unfavorable rotations can lead to steric clashes and high-energy conformations, while favorable angles result in stable structures.\n",
    "\n",
    "### **Ramachandran Plot:**\n",
    "\n",
    "The Ramachandran plot, developed by Dr. G.N. Ramachandran, is a graphical representation of the allowed and disallowed regions of Œ¶ and Œ® angles in proteins. It helps visualize which combinations of dihedral angles are energetically favorable for a given protein. The plot maps the distribution of these angles, showing regions where secondary structures like alpha helices and beta sheets are found.\n",
    "\n",
    "<center><img src=\"images/Ramachandran.png\" width=400 /></center><br><br>\n",
    "\n",
    "#### **How does it help ?**\n",
    "By visualizing the allowed regions, scientists can predict or confirm the secondary structure of a protein based on its dihedral angles. Alpha helices typically fall within a specific region, while beta sheets occupy a different area. Disallowed regions correspond to conformations with steric clashes or unfavorable energy states, making the Ramachandran plot an essential tool in validating protein models."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a90b159-7aee-4253-967e-ff9ed881d5f2",
   "metadata": {
    "id": "7a90b159-7aee-4253-967e-ff9ed881d5f2"
   },
   "source": [
    "## **Side Chain Torsion Angles (Chi Angles)**\n",
    "To better understand side chain orientations, we need to explore torsion angles, also known as chi (œá) angles. These angles represent the rotation around specific bonds within the side chain of an amino acid, which allows different conformations:\n",
    "* Chi 1 (œá1): This is the first and most common torsion angle, measuring rotation around the C-alpha (CŒ±) to C-beta (CŒ≤) bond.\n",
    "* Chi 2 (œá2), Chi 3 (œá3), Chi 4 (œá4): These angles are measured for amino acids with more complex side chains that contain additional rotatable bonds.\n",
    "\n",
    "<center><img src=\"images/torsion_angle.png\" width=800 /><br><br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b3b0542-e20f-4a9a-999d-15ca25901f92",
   "metadata": {
    "id": "6b3b0542-e20f-4a9a-999d-15ca25901f92"
   },
   "outputs": [],
   "source": [
    "#Side chain torsion angles\n",
    "from IPython.display import YouTubeVideo\n",
    "\n",
    "# Youtube\n",
    "YouTubeVideo(id='zoIBJxGOt4g', height=200, width=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09715c1e-3156-4880-a4fc-237c819ea580",
   "metadata": {
    "id": "09715c1e-3156-4880-a4fc-237c819ea580"
   },
   "source": [
    "## **Conformation and Configuration**\n",
    "Different arrangement of atoms in space in a molecule that results from rotation of a chemical group around a single bond (changing the position of the chemical groups without breaking the bond) is conformation.<br><br>\n",
    "Particular arrangement of atoms in a molecule is configuration. Example, cis and trans arrangement of atoms.\n",
    "\n",
    "<center><img src=\"images/configuration.png\" width=600 /><br><br>\n",
    "\n",
    "<center><img src=\"images/proline.png\" width=600 /><br><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfb1a2f8-0128-4694-8c1b-30e3efe14737",
   "metadata": {
    "id": "cfb1a2f8-0128-4694-8c1b-30e3efe14737"
   },
   "source": [
    "## **Homology Analysis**\n",
    "Homology analysis is an important tool in structural biology that helps us understand how proteins are related and how they evolved. By comparing the 3D shapes of similar proteins, this analysis can pinpoint key functional regions, predict where proteins might bind to other molecules, and reveal shared molecular functions across different species. By recognizing these similarities and differences, scientists can make informed guesses about the role of newly discovered proteins, study how proteins have adapted over time, or gain insight into disease mechanisms.\n",
    "\n",
    "In the activity below, we will:\n",
    "\n",
    "1. Load and view two related protein structures using PyMOL‚Äôs command line tools, coloring them differently for better clarity.\n",
    "2. Align the protein structures to compare them and calculate their similarity using the Root Mean Square Deviation (RMSD) value, which measures how alike they are in structure.\n",
    "3. Highlight interaction areas by showing residues that are close together in the aligned structures, helping us identify conserved regions or potential contact points between the proteins.\n",
    "4. Investigate active sites by identifying important residues and representing them using spheres or sticks to clearly mark their positions.\n",
    "\n",
    "This hands-on tutorial will show you how to use these tools to explore how structural similarities between proteins can provide insights into shared functions and interaction sites. Homology analysis in PyMOL is not only a valuable approach for studying evolution but also plays a crucial role in fields like drug discovery and protein engineering.\n",
    "\n",
    "Let‚Äôs get started with the activity where you‚Äôll learn how to run each command and analyze the results to understand the similarities and functional importance of related protein structures."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "468f5c2b-cf3d-42f7-8fca-44093cbbfbd2",
   "metadata": {
    "id": "468f5c2b-cf3d-42f7-8fca-44093cbbfbd2"
   },
   "source": [
    "### **Homology Analysis in Drug Discovery**  \n",
    "\n",
    "Homology analysis plays a key role in drug discovery by using evolutionary relationships between proteins to predict their structure, function, and potential drug binding sites. By comparing protein sequences and structures across different species and protein families, scientists can identify conserved regions that are often crucial for protein function and stability.  \n",
    "\n",
    "**How homology analysis helps in drug design:**  \n",
    "- Identifies conserved drug binding sites, making it easier to target functionally important regions.  \n",
    "- Builds structural models for proteins that lack experimental data, aiding in virtual screening and structure-based drug design.  \n",
    "- Predicts potential off-target effects by analyzing sequence conservation across species, improving drug selectivity.  \n",
    "\n",
    "**When is it useful?**  \n",
    "- No crystal structure is available for the target protein.  \n",
    "- Developing selective inhibitors within protein families.  \n",
    "- Investigating drug resistance mechanisms.  \n",
    "- Identifying new drug targets.  \n",
    "- Assessing drug safety across species.  \n",
    "\n",
    "**Steps in homology analysis:**  \n",
    "- Aligning protein sequences to find conserved regions.  \n",
    "- Comparing protein structures to detect similarities.  \n",
    "- Identifying conserved binding pockets and key residues.  \n",
    "- Evaluating species differences for better drug selectivity.  \n",
    "- Predicting potential off-target interactions.  \n",
    "\n",
    "In **Activity 5**, you will use various tools to perform homology analysis and explore its applications in drug discovery.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af19b3f7-72c5-4a10-a695-8d1a08a77d7d",
   "metadata": {
    "id": "af19b3f7-72c5-4a10-a695-8d1a08a77d7d"
   },
   "source": [
    "-------------\n",
    "# üìä Tutorials\n",
    "In these tutorials, we will use the PyMOL, ClusPro, and other cheminformatic tools to work through <u>**five**</u> applied activities to:\n",
    "- Measuring dihedral angles\n",
    "- Torsion angle measurements and sidechin-ligand interactions\n",
    "- Hemoglobin Ramachandran plot\n",
    "- Analyzing importance of proline\n",
    "- Homology example\n",
    "\n",
    "## Before you begin:\n",
    "- Run the PyMOL GUI by following the directions provided in the Submodule 0 notebook, provided here: [pymol_notebook](../submodule0_pymol_setup/pymol_notebook.ipynb)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c52d8966-6c4b-4f0f-993a-890b1c11157c",
   "metadata": {
    "id": "c52d8966-6c4b-4f0f-993a-890b1c11157c"
   },
   "source": [
    "## üåü **Activity 1: Visualizing and Measuring Phi (œï) and Psi (œà) Dihedral Angles**\n",
    "\n",
    "### **Objective:**  \n",
    "To visualize and measure *phi (œï)* and *psi (œà)* dihedral angles of a peptide fragment using PyMOL.\n",
    "\n",
    "### **Steps to Perform the Activity**\n",
    "\n",
    "#### **Step 1: Load the Peptide Structure**\n",
    "1. Open the PDB File using the the *GUI menu* to load the peptide structure:<br>\n",
    "     `File > Open > peptide.pdb`<br>\n",
    "   -<u>Note</u>: The instructor will provide the `peptide.pdb` file. Place it in the designated folder for easy access.\n",
    "\n",
    "2. Change the display mode to *sticks* to hide the default cartoon representation:<br>\n",
    "     `Hide > Cartoon`\n",
    "     \n",
    "   - The peptide will now be displayed in stick representation with the following *color codes*:  \n",
    "       - Green: Carbon  \n",
    "       - Blue: Nitrogen  \n",
    "       - Red: Oxygen  \n",
    "       - Yellow: Sulfur  \n",
    "\n",
    "#### **Step 2: Orient the Peptide**\n",
    "1. Display the sequence (at the top of the PyMOL interface) by navigating to:\n",
    "     `Display > Sequence`\n",
    "\n",
    "2. Select *Serine (S)* by clicking *S* in the sequence display.\n",
    "\n",
    "3. Rotate the peptide such that it follows the convention for displaying peptides:\n",
    "     - *N-terminal (Serine)* is on the *left*.\n",
    "     - *C-terminal (Aspartic acid)* is on the *right*.  \n",
    "\n",
    "#### **Step 3: Measure Dihedral Angles for Ala296 and Pro297**\n",
    "1. Set the selection mode by navigating to:<br>\n",
    "   `Mouse > Selection Mode > Atoms`\n",
    "\n",
    "2. Activate the measurement wizard by navigating to:<br>\n",
    "    `Wizard > Measurement > Distances > Dihedral`\n",
    "   - The wizard will allow you to click on atoms to measure dihedral angles.\n",
    "\n",
    "3. Measure Phi (œï) of Ala296 by clicking on the following atoms in sequence:\n",
    "     1. *C=O* of Ser295  \n",
    "     2. *N* of Ala296  \n",
    "     3. *C-alpha* of Ala296  \n",
    "     4. *C=O* of Ala296  \n",
    "   - The *œï angle* for Ala296 will be displayed as *-108¬∞*.\n",
    "\n",
    "4. Measure Psi (œà) of Ala296 by clicking on the following atoms in sequence:\n",
    "     1. *N* of Ala296  \n",
    "     2. *C-alpha* of Ala296  \n",
    "     3. *C=O* of Ala296  \n",
    "     4. *N* of Pro297  \n",
    "    - The *œà angle* for Ala296 will be displayed as *89¬∞*.\n",
    "\n",
    "5. Measure *œï* and *œà* for Pro297 using the same method outlined above.\n",
    "\n",
    "### **Understanding the Dihedral Angles**\n",
    "- <u>Phi (œï) Angle:</u> The rotation around the bond between the nitrogen (N) and the alpha carbon (CŒ±).\n",
    "- <u>Psi (œà) Angle:</u> The rotation around the bond between the alpha carbon (CŒ±) and the carbonyl carbon (C=O).\n",
    "- These angles define the secondary structure (alpha helices, beta sheets, turns) of proteins."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81303871-57fe-4158-bb32-397f920cb414",
   "metadata": {
    "id": "81303871-57fe-4158-bb32-397f920cb414"
   },
   "source": [
    "-------------------\n",
    "## üåü **Activity 2: To Plot and Analyze a Ramachandran Map for Hemoglobin**\n",
    "\n",
    "### **Objective:**  \n",
    "To generate a **Ramachandran plot** for hemoglobin and analyze the phi (œï) and psi (œà) angles of residues in the protein, categorized by their structural properties.\n",
    "\n",
    "### **Steps to Perform the Activity**\n",
    "\n",
    "#### **Step 1: Install the DynoPlot Plugin**\n",
    "1. Install the *DynoPlot Plugin* within PyMOL by following these steps:<br>\n",
    "   - Go to `Plugin > Plugin Manager > Install New Plugin` <br>\n",
    "   - Upload or install the plugin from the GitHub repository: <br>\n",
    "       + https://github.com/Pymol-Scripts/Pymol-script-repo/blob/master/plugins/dynoplot.py<br>\n",
    "   - Follow the prompts to complete installation.<br>\n",
    "\n",
    "#### **Step 2: Fetch the Hemoglobin Structure**\n",
    "1. Fetch the PDB file for Hemoglobin:<br>\n",
    "   `PyMOL> fetch 3BCQ`\n",
    "   - This command downloads and displays the hemoglobin structure in PyMOL.\n",
    "\n",
    "#### **Step 3: Generate a Ramachandran Plot**\n",
    "1. Plot the Ramachandran Map for All Residues:<br>\n",
    "   `PyMOL> rama all`\n",
    "   - This command generates a *Ramachandran plot* for all residues in the hemoglobin structure.\n",
    "   - Shapes and meanings within the plot include the following:\n",
    "     - <u>Triangles (Glycine)</u>: Glycine lacks a side chain, so it can appear in any region of the plot.\n",
    "     - <u>Squares (Proline)</u>: Proline typically has **phi = -60¬∞** due to its restricted flexibility.\n",
    "     - <u>Circles (All other residues)</u>: Represent standard amino acids, positioned according to their phi and psi values.\n",
    "\n",
    "#### **Step 4: Color the Secondary Structure**\n",
    "1. Using the PyMOL GUI, display the protein by secondary structure by following these steps:<br>\n",
    "   - Go to the menu next to `3BCQ` (on the right side).<br>\n",
    "   - Click `C (Color)* > Color by SS (Secondary Structure)`<br>\n",
    "       - This will color-code the structure based on secondary elements (e.g., helices, sheets, loops).\n",
    "\n",
    "2. Replot the Ramachandran Map with Secondary Structure Colors:<br>\n",
    "   `PyMOL> rama all`<br>\n",
    "   - The updated Ramachandran plot now uses the *secondary structure coloring scheme*.\n",
    "\n",
    "#### **Step 5: Analyze the Plot**\n",
    "1. Compare with Standard Ramachandran Plot:\n",
    "  - Analyze the distribution of residues in *allowed regions* (e.g., alpha-helical, beta-sheet regions).\n",
    "  - Compare the plot to standard Ramachandran distributions to identify structural patterns.\n",
    "2. Take not of the following <u>Key Observations</u>:\n",
    "  - Helical residues cluster in the *alpha region* (phi ‚âà -60¬∞, psi ‚âà -40¬∞).\n",
    "  - Beta-sheet residues cluster in the *beta region* (phi ‚âà -120¬∞, psi ‚âà 120¬∞).\n",
    "  - Glycine residues are widely distributed due to their flexibility.\n",
    "\n",
    "#### **Step 6: Save the Ramachandran Plot**\n",
    "1. Save the plot as a *PostScript file* for documentation:\n",
    "   `PyMOL> rama all, none, rama_plot.ps`\n",
    "   - The file `rama_plot.ps` will be saved in the working directory.\n",
    "\n",
    "### ‚ùì **Discussion Questions**\n",
    "Click through the flashcards below to explore additional questions about Ramachandran plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3de198ee-248e-4411-b656-866770dbca5b",
   "metadata": {
    "id": "3de198ee-248e-4411-b656-866770dbca5b",
    "outputId": "552f1bb6-587b-47ac-e34e-158ce60a5adc",
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Render flashcards\n",
    "from IPython.display import IFrame\n",
    "IFrame('quiz/question_flashcards_1.2.2.html', width=600, height=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6896a628-4409-42c3-a4d4-804224d1db6c",
   "metadata": {
    "id": "6896a628-4409-42c3-a4d4-804224d1db6c"
   },
   "source": [
    "---------------------\n",
    "## üåü **Activity 3: Measuring the Side Chain Torsion Angle and Understanding the Importance of Side Chain Interaction in Ligand Binding**\n",
    "\n",
    "### **Objective:**  \n",
    "To measure the *side chain torsion angle (chi)* and study the role of side chain interactions in ligand (ibuprofen) binding using PyMOL.\n",
    "\n",
    "### **Steps to Perform the Activity**\n",
    "\n",
    "#### **Step 1: Open the Pre-Prepared PyMOL Session File**\n",
    "1. Load the pre-saved PyMOL Session File:<br>\n",
    "     `PyMol> File > Open > sidechain.pse`\n",
    "    - This session file is pre-configured with the PDB structure *2BXG*, showing ibuprofen bound to human serum albumin.\n",
    "\n",
    "2. For the initial display, Ibuprofen is displayed as **thick sticks**. Parts of the human serum albumin protein involved in interaction are shown as *thin sticks*.\n",
    "\n",
    "#### **Step 2: Visualize Side Chain Interactions**\n",
    "1. Rotate and zoom into the structure to focus on the *binding pocket* where ibuprofen interacts with the protein.\n",
    "2. Observe how side chains of the protein residues interact with the ligand.\n",
    "3. Color Protein Side Chains for using different colors for interacting side chains to distinguish between them:<br>\n",
    "     `PyMol> select sidechains, resi 105+112+187+199  # Example interacting residues`<br>\n",
    "     `PyMol> color blue, sidechains`\n",
    "\n",
    "#### **Step 3: Measure Side Chain Torsion Angles (Chi)**\n",
    "1. Activate the measurement wizard by navigating to:<br>\n",
    "     `Wizard > Measurement > Distances > Dihedral`\n",
    "\n",
    "2. Select atoms for to measure the *chi angle* (side chain torsion) by clicking on the atoms in the following sequence:\n",
    "    - Alpha carbon (CŒ±)\n",
    "    - Beta carbon (CŒ≤)\n",
    "    - Gamma carbon (CŒ≥) or corresponding atom<br>\n",
    "    - Delta carbon (CŒ¥) or corresponding atom<br>\n",
    "       - The dihedral angle will be displayed on the screen.<br>\n",
    "\n",
    "3. Repeat for all key residues, measuring chi angles for all residues in contact with ibuprofen.\n",
    "\n",
    "#### **Step 4: Analyze the Side Chain Interaction**\n",
    "1. <u> Observe the Chi Angles </u>: Note the side chain orientations (chi angles) of residues in the binding pocket.\n",
    "   Analyze whether the angles optimize interaction with ibuprofen.\n",
    "\n",
    "2. Observer the following <u>Key Points</u>:**\n",
    "   - Hydrogen bonding.\n",
    "   - Hydrophobic interactions.\n",
    "   - Steric complementarity between ibuprofen and the protein side chains.\n",
    "\n",
    "#### ‚ùì **Additional Discussion Questions**\n",
    "Click through the flashcards below to explore additional questions about sidechain angles and ligand binding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b38f11f-f45b-400f-a2d6-d52ecd4f54a0",
   "metadata": {
    "id": "0b38f11f-f45b-400f-a2d6-d52ecd4f54a0",
    "outputId": "07edaa2c-5b58-4f8f-a815-bbb2e1098e26",
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Render flashcards\n",
    "from IPython.display import IFrame\n",
    "IFrame('quiz/question_flashcards_1.2.4.html', width=600, height=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23860ff1-acb2-4130-9f59-c86b00fad160",
   "metadata": {
    "id": "23860ff1-acb2-4130-9f59-c86b00fad160"
   },
   "source": [
    "## üåü **Activity 4: Studying the Proline Conformation**\n",
    "\n",
    "### **Objective:**  \n",
    "To study the unique conformation of <u>proline</u> residues in a protein using PyMOL.\n",
    "\n",
    "### **Steps to Perform the Activity**\n",
    "\n",
    "#### **Step 1: Load the Protein Structure**\n",
    "1. Fetch and display the PDB file using the following PyMOL command:<br>\n",
    "     `PyMol> fetch 5GV0`\n",
    "\n",
    "2. Modify the display to hide the default cartoon representation:<br>\n",
    "     `PyMol> hide cartoon`\n",
    "     \n",
    "3. Show the structure in *stick representation*:<br>\n",
    "     `PyMol> show sticks`\n",
    "     \n",
    "#### **Step 2: Select Proline Residues**\n",
    "1. Use the following command to select all prolines in the structure:<br>\n",
    "     `PyMol> select ////PRO/ca`\n",
    "   - This selects only the *alpha carbon (CŒ±)* atoms of proline residues.\n",
    "\n",
    "2. Color the selected prolines for better visualization:<br>\n",
    "     `PyMol> color red, sele`\n",
    "\n",
    "#### **Step 3: Analyze Proline's Conformation**\n",
    "1. <u>Observe the Conformation</u>: Rotate and zoom in to examine the cyclic structure of proline. Proline has a rigid structure due to its cyclic side chain, which restricts its *phi (œï)* dihedral angle to around *-60¬∞*.\n",
    "2. Note the following <u>Key Characteristics</u>:\n",
    "   - Proline's rigidity introduces *kinks* in secondary structures.\n",
    "   - It often occurs in *turns* or *loops* rather than helices or sheets."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aedf7ae6-245f-4c4d-a3d1-058e881f97d2",
   "metadata": {
    "id": "aedf7ae6-245f-4c4d-a3d1-058e881f97d2"
   },
   "source": [
    "--------------------\n",
    "## üåü **Activity 5: Homology Analysis PyMol Example**\n",
    "Homology analysis is an important tool in structural biology that helps us understand how proteins are related and how they evolved. By comparing the 3D shapes of similar proteins, this analysis can pinpoint key functional regions, predict where proteins might bind to other molecules, and reveal shared molecular functions across different species. By recognizing these similarities and differences, scientists can make informed guesses about the role of newly discovered proteins, study how proteins have adapted over time, or gain insight into disease mechanisms. In this tutorial, we will explore various aspects of homology analysis using PyMOL, from structural alignment to measuring specific distances and angles between important regions.\n",
    "\n",
    "### **Objectives:**\n",
    "- Load and view two related protein structures using PyMOL‚Äôs command line tools, coloring them differently for better clarity.\n",
    "- Align the protein structures to compare them and calculate their similarity using the Root Mean Square Deviation (RMSD) value, which measures how alike they are in structure.\n",
    "- Highlight interaction areas by showing residues that are close together in the aligned structures, helping us identify conserved regions or potential contact points between the proteins.\n",
    "- Investigate active sites by identifying important residues and representing them using spheres or sticks to clearly mark their positions.\n",
    "\n",
    "### **Steps to Perform the Activity**\n",
    "\n",
    "#### **Step 1: Load the Structures**\n",
    "1. Run the following PyMOL commands to load the protein structures: <br>\n",
    "`PyMol> fetch 1A3N, async=0  # Load the first structure (e.g., protein A)` <br>\n",
    "`PyMol>fetch 1BVC, async=0  # Load the second structure (e.g., protein B)`<br>\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 1 Results</summary>\n",
    "  <center><img src=\"images/step1.png\" width=400 /></center><br><br>\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 2: Set Initial Representation for Better Clarity**\n",
    "1. Run the following PyMOL commands to set appropriate display parameters: <br>\n",
    "`PyMol> hide everything, all  # Hide all elements initially for clarity` <br>\n",
    "`PyMol> show cartoon, all  # Show cartoon representation for all proteins` <br>\n",
    "`PyMol> color cyan, 1A3N  # Color protein A cyan` <br>\n",
    "`PyMol> color magenta, 1BVC  # Color protein B magenta` <br>\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 2 Results</summary>\n",
    "  <center><img src=\"images/step2.png\" width=400 /></center><br><br>\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 3: Align the Structures**\n",
    "1. Align protein B (1BVC) to protein A (1A3N). This step aligns the backbone atoms of the structures to assess their homology.<br>\n",
    "`PyMol> align 1BVC, 1A3N  # Align protein B (1BVC) to protein A (1A3N)`\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 3 Results</summary>\n",
    "  <center><img src=\"images/step3.png\" width=400 /></center><br><br>\n",
    "    \n",
    "  ##### RESULT:\n",
    "  - Match: assigning 464 x 1029 pairwise scores.\n",
    "  - MatchAlign: aligning residues (464 vs 1029)...\n",
    "  - MatchAlign: score 114.000\n",
    "  - ExecutiveAlign: 862 atoms aligned.\n",
    "  - ExecutiveRMS: 33 atoms rejected during cycle 1 (RMSD=2.56).\n",
    "  - ExecutiveRMS: 26 atoms rejected during cycle 2 (RMSD=1.71).\n",
    "  - ExecutiveRMS: 19 atoms rejected during cycle 3 (RMSD=1.57).\n",
    "  - ExecutiveRMS: 10 atoms rejected during cycle 4 (RMSD=1.50).\n",
    "  - ExecutiveRMS: 5 atoms rejected during cycle 5 (RMSD=1.47).\n",
    "  - Executive: RMSD =    1.456 (769 to 769 atoms)\n",
    "\n",
    "    The two structures are aligned. The output will include the RMSD (Root Mean Square Deviation) value, which measures how similar the two structures are.\n",
    "</details>\n",
    "\n",
    "#### **Step 4: Refined Alignment**\n",
    "1. Perform a more refined alignment using the `super` command, which can be more accurate than `align`:<br>\n",
    "    `PyMol> super 1BVC, 1A3N`\n",
    "    - The superposition is refined further, and RMSD values are recalculated, often resulting in a lower RMSD.\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 4 Results</summary>\n",
    "    \n",
    "  ##### RESULT:\n",
    "  - MatchAlign: aligning residues (153 vs 574)...\n",
    "  - MatchAlign: score 540.114\n",
    "  - ExecutiveAlign: 862 atoms aligned.\n",
    "  - ExecutiveRMS: 32 atoms rejected during cycle 1 (RMSD=1.93).\n",
    "  - ExecutiveRMS: 24 atoms rejected during cycle 2 (RMSD=1.58).\n",
    "  - ExecutiveRMS: 14 atoms rejected during cycle 3 (RMSD=1.49).\n",
    "  - ExecutiveRMS: 12 atoms rejected during cycle 4 (RMSD=1.45).\n",
    "  - ExecutiveRMS: 11 atoms rejected during cycle 5 (RMSD=1.41).\n",
    "  - Executive: RMSD =    1.379 (769 to 769 atoms)\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 5: Calculate RMSD for Backbone Atoms**\n",
    "1. We calculate RMSD specifically for CA (alpha carbon) atoms of the structures to evaluate backbone alignment. <br>\n",
    "`PyMol> rms_cur 1BVC and name CA, 1A3N and name CA`\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 5 Results</summary>\n",
    "    \n",
    "  ##### RESULT:\n",
    "  - Executive: RMSD =  5.743 (25 to 25 atoms)\n",
    "\n",
    "    The RMSD value between the two structures is 27.242 √Ö for 25 CA atoms, which indicates their alignment. A higher RMSD usually means less similarity between the aligned structures.\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 6: Perform Alignment Without Refinement Cycles**\n",
    "1. Use the `align` command with zero refinement cycles to see the initial alignment before refinement. <br>\n",
    "`align 1BVC, 1A3N, cycles=0`\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 6 Results</summary>\n",
    "    \n",
    "  <center> <img src=\"images/step6.png\" width=400 /></center><br><br>\n",
    "    \n",
    "  ##### RESULT:\n",
    "  - Match: read scoring matrix.\n",
    "  - Match: assigning 464 x 1029 pairwise scores.\n",
    "  - MatchAlign: aligning residues (464 vs 1029)...\n",
    "  - MatchAlign: score 114.000\n",
    "  - ExecutiveAlign: 862 atoms aligned.\n",
    "  - Executive: RMSD =    2.563 (862 to 862 atoms)\n",
    "\n",
    "    The alignment score is 114. The alignment without refinement cycles provides an RMSD of 2.563 √Ö for 862 atoms.\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 7: Partial Region Alignment**\n",
    "1. Align residues 50-100 of both structures to focus on a specific region.<br>\n",
    "`PyMol> align 1BVC and resi 50-100, 1A3N and resi 50-100`\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 7 Results</summary>\n",
    "    \n",
    "  <img src=\"images/step7.png\" width=600 /><br><br>\n",
    "    \n",
    "  ##### RESULT:\n",
    "  - Match: read scoring matrix.\n",
    "  - Match: assigning 51 x 204 pairwise scores.\n",
    "  - MatchAlign: aligning residues (51 vs 204)...\n",
    "  - MatchAlign: score 38.000\n",
    "  - ExecutiveAlign: 296 atoms aligned.\n",
    "  - ExecutiveRMS: 26 atoms rejected during cycle 1 (RMSD=5.96).\n",
    "  - ExecutiveRMS: 11 atoms rejected during cycle 2 (RMSD=3.19).\n",
    "  - ExecutiveRMS: 12 atoms rejected during cycle 3 (RMSD=1.53).\n",
    "  - ExecutiveRMS: 16 atoms rejected during cycle 4 (RMSD=1.23).\n",
    "  - ExecutiveRMS: 11 atoms rejected during cycle 5 (RMSD=1.04).\n",
    "  - Executive: RMSD =    0.931 (220 to 220 atoms)\n",
    "\n",
    "    The alignment between residues 50-100 yields an RMSD of 0.931 √Ö, indicating a good alignment for that specific region.\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 8: Highlight Close Interacting Regions**\n",
    "1. Highlight atoms within 2 √Ö of each structure to observe close contacts.<br>\n",
    "`PyMol> color red, (1BVC within 2 of 1A3N)  # Color atoms in 1BVC that are within 2 √Ö of 1A3N in red`<br>\n",
    "`PyMol> color blue, (1A3N within 2 of 1BVC)  # Color atoms in 1A3N that are within 2 √Ö of 1BVC in blue`<br>\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 8 Results</summary>\n",
    "    \n",
    "  <img src=\"images/step8.png\" width=600 /><br><br>\n",
    "    \n",
    "  ##### RESULT:\n",
    "  - Executive: Colored 1171 atoms\n",
    "  - Executive: Colored 1151 atoms\n",
    "\n",
    "    This highlights atoms involved in close contacts between the two proteins, providing insight into possible interaction sites.\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 9: Measure Specific Distance and Angles**\n",
    "1. Measure a specific distance and an angle in the aligned structures to understand spatial relationships.<br>\n",
    "`PyMol> distance my_distance, 1BVC/100/CA, 1A3N/100/CA`<br>\n",
    "`PyMol> angle my_angle, 1BVC/50/N, 1BVC/51/CA, 1BVC/52/C`\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 9 Results</summary>\n",
    "    \n",
    "  ##### RESULT:\n",
    "    PyMOL will show a visual distance measurement and a calculated angle between the specified atoms.\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 10: Select and Visualize an Active Site**\n",
    "1. Select a hypothetical active site in the structure and display it in a different style.<br>\n",
    "`PyMol> select active_site, resi 50+58+62 and 1A3N`<br>\n",
    "`PyMol> show spheres, active_site`<br>\n",
    "`PyMol> color yellow, active_site`<br>\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 10 Results</summary>\n",
    "    \n",
    "  <img src=\"images/step10.png\" width=600 /><br><br>\n",
    "    \n",
    "  ##### RESULT:\n",
    "    - Selector: selection \"active_site\" defined with 92 atoms\n",
    "    - Executive: Colored 92 atoms\n",
    "    \n",
    "    The active site is highlighted with spheres in yellow, making it stand out for better visualization.\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 11: Show Interacting Residues Between the Structures**\n",
    "1. Show interacting residues from one structure that are within a certain distance from the other.<br>\n",
    "`PyMol> show sticks, (1BVC within 3 of 1A3N and not 1A3N)  # Show interacting residues from one structure not found in the other`\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 11 Results</summary>\n",
    "    \n",
    "  ##### RESULT:\n",
    "    Interacting residues are displayed in stick representation, which can be useful for studying potential binding interactions.\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Step 12: Surface Representation of Ligand Binding Pocket**\n",
    "1. Use the surface representation with adjusted transparency to highlight the ligand binding pocket.<br>\n",
    "`PyMol> show surface, 1A3N`\n",
    "`PyMol> set transparency, 0.5, 1A3N  # Use surface representation and adjust transparency to show the ligand binding pocket clearly`\n",
    "\n",
    "<details>\n",
    "  <summary>Click to see Step 12 Results</summary>\n",
    "  \n",
    "  <img src=\"images/step12.png\" width=600 /><br><br>\n",
    "    \n",
    "  ##### RESULT:\n",
    "    - Setting: transparency set to 0.50000 in object \"1A3N\"\n",
    "    \n",
    "    The surface of 1A3N is shown with a transparency of 50%, making it easy to visualize the underlying binding pocket while also seeing through to the cartoon representation.\n",
    "\n",
    "</details>\n",
    "\n",
    "#### **Additional Analyses:**\n",
    "You can perform additional analyses on the obtained results, such as:\n",
    "\n",
    "1. <u>Active Site Characterization</u>: Using the active site selections (active_site), you can dive deeper into the geometry of residues involved in binding or catalysis. Visualizing these residues in different representations (e.g., spheres or sticks) can reveal key structural features important for ligand binding, which is particularly useful in drug design.\n",
    "\n",
    "2. <u>Protein-Ligand Interaction Analysis</u>: With the distances (my_distance) and angles (my_angle) calculated between atoms, you can understand the stability and orientation of a ligand within a binding site. Such information is critical when optimizing a ligand to enhance binding affinity or designing inhibitors that fit precisely.\n",
    "\n",
    "3. <u>Comparative Structure Analysis</u>: The alignment steps and the RMSD values we derived allow us to understand how similar two protein structures are. This kind of analysis is essential to determine conserved structural features among homologous proteins, potentially hinting at similar biological functions or evolutionary relationships.\n",
    "\n",
    "#### **Key Takeaways:**\n",
    "- Highlighting close contacts and coloring interacting residues provides a way to focus on the critical regions that mediate protein-protein or protein-ligand interactions. This helps in understanding how changes in these regions (e.g., mutations) might affect overall protein stability or function.\n",
    "- By using the PyMOL commands that we discussed, you can extend this analysis further, whether for drug discovery, protein engineering, or understanding the effects of mutations on protein activity. These approaches offer detailed insights into the structural relationships between proteins, their functional domains, and how they can interact with ligands or other biomolecules."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb999a47-e596-4c43-803e-d559b10dd93f",
   "metadata": {
    "id": "eb999a47-e596-4c43-803e-d559b10dd93f"
   },
   "source": [
    "---------------\n",
    "## Conclusions\n",
    "Users should be able to understand the anatomy of protein structures in detail, including measuring backbone and side-chain torsion angles, Phi, psi and Chi using PDB file os any protein 3D structure and analyze the Ramachandran plot for allowed phi and psi values in the protein structure. They should also be able to analyse the protein sequence information and compare the sequences of different proteins for homologous proteins using Python code.\n",
    "## Clean Up\n",
    "<div class=\"alert alert-block alert-warning\"> <b>Attention:</b> Remember to shutdown VM and delete any relevant resources</a>. </div>"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "environment": {
   "kernel": "conda-base-py",
   "name": "workbench-notebooks.m128",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/workbench-notebooks:m128"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel) (Local)",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
